README.md
<a href="http://www.ffuenf.de" title="ffuenf - code • design • e-commerce"><img src="https://github.com/ffuenf/Ffuenf_Common/blob/master/skin/adminhtml/default/default/ffuenf/ffuenf.png" alt="ffuenf - code • design • e-commerce" /></a>

Ffuenf_DevTools
===============
[![GitHub tag](https://img.shields.io/github/tag/ffuenf/Ffuenf_DevTools.svg)](https://github.com/ffuenf/Ffuenf_DevTools)
[![Build Status](https://img.shields.io/travis/ffuenf/Ffuenf_DevTools.svg)](https://travis-ci.org/ffuenf/Ffuenf_DevTools)
[![Code Quality](https://scrutinizer-ci.com/g/ffuenf/Ffuenf_DevTools/badges/quality-score.png)](https://scrutinizer-ci.com/g/ffuenf/Ffuenf_DevTools)
[![Code Coverage](https://scrutinizer-ci.com/g/ffuenf/Ffuenf_DevTools/badges/coverage.png)](https://scrutinizer-ci.com/g/ffuenf/Ffuenf_DevTools)
[![PayPal Donate](https://img.shields.io/badge/paypal-donate-blue.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=J2PQS2WLT2Y8W&item_name=Magento%20Extension%3a%20Ffuenf_DevTools&item_number=Ffuenf_DevTools&currency_code=EUR)

This is a extension for Magento Community Edition that adds access to commonly used development tasks in the backend.

Replacements
------------

This extension inherits the functionality of the following extensions:

* [N98_LayoutHelper](https://github.com/netz98/N98_LayoutHelper)
* [N98_CustomLayoutHandles](https://github.com/netz98/N98_CustomLayoutHandles)
* [Aoe_LayoutConditions](https://github.com/AOEpeople/Aoe_LayoutConditions)
* [MagePsycho_Xmllinks](https://github.com/MagePsycho/MagePsycho_Xmllinks)
* [MageMechanic_PhpSeven](https://github.com/MageMechanic/PhpSeven)
* [Philwinkle_AppliedPatches](https://github.com/philwinkle/Philwinkle_AppliedPatches)
* [Yireo_EmailOverride](https://github.com/yireo/Yireo_EmailOverride)
* [Dh_EmptyHandles](https://github.com/drewhunter/EmptyHandles)
* [Mexbs_Urlrewrite](http://www.mexbs.com/product/mass-delete-url-rewrites-free-magento-extension/)
* [Inchoo_Xternal](http://inchoo.net/magento/how-to-add-external-javascript-css-file-to-magento/)

Platform
--------

The following versions are supported and tested:

* Magento Community Edition 1.9.2.4
* Magento Community Edition 1.9.1.1
* Magento Community Edition 1.8.1.0
* Magento Community Edition 1.7.0.2
* Magento Community Edition 1.6.2.0

Other versions are assumed to work.

Requirements
------------

|                                                                     | PHP 5.5           | PHP 5.6           | PHP 7.0           |
| ------------------------------------------------------------------- | ----------------- | ----------------- | ----------------- |
| [EOL](https://secure.php.net/supported-versions.php) / STABLE / RC  | STABLE            | **STABLE**        | **STABLE**        |
| automated tests on [travis]                                         | **required pass** | **required pass** | **required pass** |

Magento Community Edition officially supports PHP 5.4 and PHP 5.5.

Non-official compatibility to PHP 5.6 may be reached by following the tips on [Use of iconv.internal_encoding is deprecated](https://magento.stackexchange.com/questions/34015/magento-1-9-php-5-6-use-of-iconv-internal-encoding-is-deprecated).
Non-official compatibility to PHP 7.0 may be reached by using [Inchoo_PHP7](https://github.com/Inchoo/Inchoo_PHP7).

Installation
------------

Use [modman](https://github.com/colinmollenhour/modman) to install:
```
modman init
modman clone https://github.com/ffuenf/Ffuenf_Common
modman clone https://github.com/ffuenf/Ffuenf_DevTools
```

Deinstallation
--------------

#### via [Ffuenf_MageTrashApp](https://github.com/ffuenf/Ffuenf_MageTrashApp)

An additional module called [Ffuenf_MageTrashApp](https://github.com/ffuenf/Ffuenf_MageTrashApp) has been installed to help you to uninstall this extension in a clean way.
If it is not yet installed, please install it from [Ffuenf_MageTrashApp](https://github.com/ffuenf/Ffuenf_MageTrashApp)
If it is installed, go to your backend menu System > Configuration > Advanced > MageTrashApp, then click on the tab "Extension Installed", select the drop down option "Uninstall" of this extension and press "Save Config" button to uninstall
If you use this extension, you don't need to make any queries in your database as explained below in case of manually uninstallation.

#### via [modman](https://github.com/colinmollenhour/modman)

Use [modman](https://github.com/colinmollenhour/modman) to clear all files and symlinks:
```
modman clean Ffuenf_DevTools
```
see `sql/ffuenf_devtools_setup/uninstall.sql` to clear all entries of this extension from your database.

Development
-----------
1. Fork the repository from GitHub.
2. Clone your fork to your local machine:

        $ git clone https://github.com/USER/Ffuenf_DevTools

3. Create a git branch

        $ git checkout -b my_bug_fix

4. Make your changes/patches/fixes, committing appropriately
5. Push your changes to GitHub
6. Open a Pull Request

License and Author
------------------

- Author:: Achim Rosenhagen (<a.rosenhagen@ffuenf.de>)
- Copyright:: 2017, ffuenf

The MIT License (MIT)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

CHANGELOG.md
# CHANGELOG for Ffuenf_DevTools

This file is used to list changes made in each version of Ffuenf_DevTools.

## 2.2.0 (September 4, 2017)

* add layout handles for empty category empty search and empty cart pages
* add notes about inherited extensions
* add soaptest
* add log configuration
* remove magento composer repo
* several minor fixes

## 2.1.9 (March 12, 2016)

* update travis / MageTestStand
* [docs] minor updates
* fix helper
* re-organize menu / navigation
* fix date input

## 2.1.8 (February 25, 2016)

* update travis build-matrix
* [docs] copyright notice

## 2.1.7 (December 20, 2015)

* fix date input in backend configuration
* add date type config for resave-products shell script (created_at / updated_at)
* fix dependency to [Ffuenf_Common](https://github.com/ffuenf/Ffuenf_Common)
* code-style
* update travis build-matrix

## 2.1.6 (November 22, 2015)

* add uninstallation via [Ffuenf_MageTrashApp](https://github.com/ffuenf/Ffuenf_MageTrashApp)

## 2.1.5 (November 21, 2015)

* add dependency to [Ffuenf_Common](https://github.com/ffuenf/Ffuenf_Common)
* common extension structure
* add more tests
* add note about inherited extensions

## 2.1.4 (November 8, 2015)

* switch back to official builds for versions > 1.9.0.0 in travis tests
* notes about compatibility

## 2.1.3 (November 6, 2015)

* switch completely to OpenMage mirrors in travis tests
* cache helper in model
* code style
* relocate admin routes (compatibility to SUPEE-6788)

## 2.1.2 (November 4, 2015)

* add resave-products shell script (timeframe configurable via backend settings)
* fix wrong helpers
* fix compatibility to [SUPEE-6788](https://info2.magento.com/rs/318-XBX-392/images/SUPEE-6788-Technical%20Details.pdf?mkt_tok=3RkMMJWWfF9wsRolv6jPZKXonjHpfsX66uskXqOxlMI%2F0ER3fOvrPUfGjI4CSMFjI%2BSLDwEYGJlv6SgFSbHNMbhiwrgOUhM%3D)

## 2.1.1 (November 1, 2015)

* add more layout xml conditions (credits to [magehandles](https://github.com/aleron75/magehandles) by Alessandro Ronchi)

## 2.1.0 (October 30, 2015)

* add layout xml conditions (credits to [Aoe_LayoutConditions](https://github.com/fbrnc/Aoe_LayoutConditions) by Fabrizio Branca)
* conditionally enable headerbar in backend (and make css dependent)
* add test whether extension uses the old-style admin routing.

## 2.0.3 (October 29, 2015)

* update to notice of applied patches
* update to possibility of overriding email templates in your own Magento theme

## 2.0.2 (October 22, 2015)

* compatibility to [SUPEE-6788](https://info2.magento.com/rs/318-XBX-392/images/SUPEE-6788-Technical%20Details.pdf?mkt_tok=3RkMMJWWfF9wsRolv6jPZKXonjHpfsX66uskXqOxlMI%2F0ER3fOvrPUfGjI4CSMFjI%2BSLDwEYGJlv6SgFSbHNMbhiwrgOUhM%3D)

## 2.0.1 (July 18, 2015)

* add possibility to mass delete URL-Rewrites

## 2.0.0 (May 25, 2015)

* add environment header in backend (credits to [Magento-AdminHeaderTheme](https://github.com/Zookal/Magento-AdminHeaderTheme) by Cyrill Schumacher)
* add git infos to header in backend
* add setting for cutom admin theme
* add notice of applied patches (credits to [Philwinkle_AppliedPatches](https://github.com/philwinkle/Philwinkle_AppliedPatches) by Phillip Jackson)
* add possibility of overriding email templates in your own Magento theme (credits to [Yireo_EmailOverride](https://github.com/yireo/Yireo_EmailOverride) by Yireo)

## 1.0.1 (May 21, 2015)

* remove composer dependencies
* add codeclimate / scrutinizer-ci / travis

## 1.0.0 (January 8, 2015)

* initial public release
* add travis-ci
* add basic test


app/code/community/Ffuenf/DevTools/Helper/Data.php
<?php
/**
 * Ffuenf_DevTools extension.
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the MIT License
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/mit-license.php
 *
 * @category   Ffuenf
 *
 * @author     Achim Rosenhagen <a.rosenhagen@ffuenf.de>
 * @copyright  Copyright (c) 2017 ffuenf (http://www.ffuenf.de)
 * @license    http://opensource.org/licenses/mit-license.php MIT License
 */

class Ffuenf_DevTools_Helper_Data extends Ffuenf_Common_Helper_Core
{
    const CONFIG_EXTENSION_ACTIVE = 'ffuenf_devtools/general/enable';
    const CONFIG_EXTENSION_LOG_SYSTEM_ACTIVE = 'ffuenf_devtools/log/enable';
    const CONFIG_EXTENSION_LOG_PROFILE_ACTIVE = 'ffuenf_devtools/log/profile_enable';
    const CONFIG_EXTENSION_LOG_EXCEPTION_ACTIVE = 'ffuenf_devtools/log/exception_enable';
    const CONFIG_EXTENSION_MAGERUNPATH = 'ffuenf_devtools/magerun/path';
    const CONFIG_EXTENSION_BACKUP_DUMPDATABASESCRIPTPATH = 'ffuenf_devtools/backup/dump_database_script_path';
    const CONFIG_EXTENSION_RESAVEPRODUCTS_TIMEFRAMEFROM = 'ffuenf_devtools/resaveproducts/timeframefrom';
    const CONFIG_EXTENSION_RESAVEPRODUCTS_TIMEFRAMETO = 'ffuenf_devtools/resaveproducts/timeframeto';
    const CONFIG_EXTENSION_RESAVEPRODUCTS_TIMEFRAMETYPE = 'ffuenf_devtools/resaveproducts/timeframetype';
    const CONFIG_EXTENSION_HEADERBAR_ACTIVE = 'ffuenf_devtools/headerbar/enable';
    const CONFIG_EXTENSION_HEADERBAR_BRANCH = 'ffuenf_devtools/headerbar/branch';
    const CONFIG_EXTENSION_HEADERBAR_COMMIT = 'ffuenf_devtools/headerbar/commit';
    const CONFIG_EXTENSION_HEADERBAR_TAG = 'ffuenf_devtools/headerbar/tag';

    private $_domain = null;
    private $_className = null;

    /**
     * Variable for if the extension is active.
     *
     * @var bool
     */
    protected $_bExtensionActive;

    /**
     * Variable for if magerun is available.
     *
     * @var bool
     */
    protected $_bMagerunAvailable;

    /**
     * Variable for the magerun system path.
     *
     * @var string
     */
    protected $_sMagerunPath;

    /**
     * Variable for the path to the dump_database shell script.
     *
     * @var string
     */
    protected $_sDumpDatabaseScriptPath;

    /**
     * Variable for the timeframe of resaving products
     *
     * @var string
     */
    protected $_sResaveProductsTimeframeFrom;

    /**
     * Variable for the timeframe of resaving products
     *
     * @var string
     */
    protected $_sResaveProductsTimeframeTo;

    /**
     * Variable for the timeframe type of resaving products
     *
     * @var string
     */
    protected $_sResaveProductsTimeframeType;

    /**
     * Variable for the headerbar.
     *
     * @var bool
     */
    protected $_bHeaderbarActive;

    /**
     * Variable for the current git branch.
     *
     * @var string
     */
    protected $_sHeaderbarBranch;

    /**
     * Variable for the current git commit.
     *
     * @var string
     */
    protected $_sHeaderbarCommit;

    /**
     * Variable for the current git tag.
     *
     * @var string
     */
    protected $_sHeaderbarTag;

    /**
     * Check to see if the extension is active.
     *
     * @return bool
     */
    public function isExtensionActive()
    {
        return $this->getStoreFlag(self::CONFIG_EXTENSION_ACTIVE, '_bExtensionActive');
    }

    /**
     * Check to see if logging is active.
     *
     * @return bool
     */
    public function isLogActive()
    {
        return Mage::getStoreConfigFlag(self::CONFIG_EXTENSION_LOG_SYSTEM_ACTIVE);
    }

    /**
     * Check to see if profile logging is active.
     *
     * @return bool
     */
    public function isLogProfileActive()
    {
        return Mage::getStoreConfigFlag(self::CONFIG_EXTENSION_LOG_PROFILE_ACTIVE);
    }

    /**
     * Check to see if exception logging is active.
     *
     * @return bool
     */
    public function isLogExceptionActive()
    {
        return Mage::getStoreConfigFlag(self::CONFIG_EXTENSION_LOG_EXCEPTION_ACTIVE);
    }

    /**
     * Checks if magerun is available.
     *
     * @return bool
     */
    public function isMagerunAvailable()
    {
        $output = $this->runMagerun(array('--version'));
        if (!isset($output[0]) || strpos($output[0], 'n98-magerun version') === false) {
            $this->_bMagerunAvailable = true;
        } else {
            $this->_bMagerunAvailable = false;
        }

        return $this->_bMagerunAvailable;
    }

    /**
     * Checks if magerun is present and returns the version number.
     *
     * @return string
     *
     * @throws Mage_Core_Exception
     */
    public function checkMagerun()
    {
        $output = $this->runMagerun(array('--version'));
        if (!$this->getMagerunPath()) {
            Mage::throwException('No valid magerun found');
        }
        $matches = array();
        preg_match('/(\d+\.\d+\.\d)/', $output[0], $matches);

        return $matches[1];
    }

    /**
     * Get magerun path.
     *
     * @return string
     *
     * @throws Mage_Core_Exception
     */
    public function getMagerunPath()
    {
        $baseDir = Mage::getBaseDir();
        $path = $baseDir . DS . $this->getStoreConfig(self::CONFIG_EXTENSION_MAGERUNPATH, '_sMagerunPath');
        if (!is_file($path)) {
            Mage::throwException('Could not find magerun at ' . $path);
        }

        return $path;
    }

    /**
     * Run magerun command.
     *
     * @param array<string>
     */
    public function runMagerun($options = array())
    {
        array_unshift($options, '--root-dir=' . Mage::getBaseDir());
        array_unshift($options, '--no-interaction');
        array_unshift($options, '--no-ansi');
        $output = array();
        exec('php -d ' . $this->getMagerunPath() . ' ' . implode(' ', $options), $output);

        return $output;
    }

    /**
     * Get database_dump script path.
     *
     * @return string
     *
     * @throws Mage_Core_Exception
     */
    public function getDatabaseDumpScriptPath()
    {
        return $this->getStoreConfig(self::CONFIG_EXTENSION_BACKUP_DUMPDATABASESCRIPTPATH, '_sDumpDatabaseScriptPath');
    }

    /**
     * @param string  $localeCode
     * @param string  $fileName
     * @param string|int|Mage_Core_Model_Store|null $store  (optional)
     *
     * @return string|null
     */
    public function getLocaleOverrideFile($localeCode, $fileName, $store = null)
    {
        $paths = $this->getLocalePaths($store);
        $localCodes = $localeCode === 'en_US' ? array($localeCode) : array($localeCode, 'en_US');
        foreach ($localCodes as $localeCode) {
            foreach ($paths as $path) {
                $filePath = $path . DS . $localeCode . DS . $fileName;
                if (!empty($filePath) && file_exists($filePath)) {
                    return $filePath;
                }
            }
        }
        return null;
    }

    /**
     * @param string|int|Mage_Core_Model_Store|null $store (optional)
     *
     * @return array
     */
    public function getLocalePaths($store = null)
    {
        $paths = array();
        $design = $this->getDesign($store);
        $paths[] = Mage::getBaseDir('design') . DS . 'frontend' . DS . $design['package'] . DS . $design['theme'].DS.'locale';
        // Check for fallback support
        if ($this->supportsDesignFallback()) {
            $fallbackModel = Mage::getModel('core/design_fallback');
            if (!empty($fallbackModel)) {
                $fallbackSchemes = $fallbackModel->getFallbackScheme('frontend', $design['package'], $design['theme']);
                if (!empty($fallbackSchemes)) {
                    foreach ($fallbackSchemes as $scheme) {
                        if (!isset($scheme['_package']) || !isset($scheme['_theme'])) {
                            continue;
                        }
                        $paths[] = Mage::getBaseDir('design') . DS . 'frontend' . DS . $scheme['_package'] . DS . $scheme['_theme'] . DS . 'locale';
                    }
                }
            }
        }
        $paths[] = Mage::getBaseDir('design') . DS . 'frontend' . DS . $design['package'] . DS . 'default' . DS . 'locale';
        $paths[] = Mage::getBaseDir('design') . DS . 'frontend' . DS . 'default' . DS . 'default' . DS . 'locale';
        $paths[] = Mage::getBaseDir('design') . DS . 'frontend' . DS . 'base' . DS . 'default' . DS . 'locale';
        $paths[] = Mage::getBaseDir('locale');
        return $paths;
    }

    /**
     * @param string|int|Mage_Core_Model_Store|null $store (optional)
     *
     * @return array
     */
    public function getDesign($store = null)
    {
        if (empty($store)) {
            $store = Mage::registry('emailoverride.store');
        }
        $packageName = null;
        $theme = null;
        if (Mage::app()->getStore()->isAdmin() == false) {
            $package = Mage::getSingleton('core/design_package');
            $originalArea = $package->getArea();
            $originalStore = $package->getStore();
            if (!empty($store)) {
                $package->setStore($store);
            }
            $package->setArea('frontend');
            $packageName = $package->getPackageName();
            $theme = $package->getTheme('default');
            $package->setArea($originalArea);
            $package->setStore($originalStore);
        }
        if (empty($packageName) || in_array($theme, array('base', 'default'))) {
            $packageName = Mage::getStoreConfig('design/package/name', $store);
        }
        if (empty($theme) || in_array($theme, array('default'))) {
            $theme = Mage::getStoreConfig('design/theme/locale', $store);
        }
        if (empty($theme)) {
            $theme = Mage::getStoreConfig('design/theme/default', $store);
        }
        if (empty($packageName)) {
            $packageName = 'default';
        }
        if (empty($theme)) {
            $theme = 'default';
        }
        return array(
            'package' => $packageName,
            'theme' => $theme,
        );
    }

    /**
     * @return string
     */
    public function getPatches()
    {
        return Mage::getModel('ffuenf_devtools/patches')->getPatches();
    }

    /**
     * Check to see if the headerbar is active.
     *
     * @return bool
     */
    public function isHeaderbarActive()
    {
        return $this->getStoreFlag(self::CONFIG_EXTENSION_HEADERBAR_ACTIVE, '_bHeaderbarActive');
    }

    /**
     * @return string
     */
    public function getTag()
    {
        $tag = '';
        exec('git describe --abbrev=0 --tags', $tag);
        $this->_sHeaderbarTag = array_shift($tag);

        return $this->getStoreConfig(self::CONFIG_EXTENSION_HEADERBAR_TAG, '_sHeaderbarTag');
    }

    /**
     * @return string
     */
    public function getBranch()
    {
        exec('git branch', $lines);
        if (is_array($lines)) {
            foreach ($lines as $line) {
                if (strpos($line, '*') === 0) {
                    $branch = ltrim($line, '* ');
                    $this->_sHeaderbarBranch = $branch;
                    break;
                }
            }
        }

        return $this->getStoreConfig(self::CONFIG_EXTENSION_HEADERBAR_BRANCH, '_sHeaderbarBranch');
    }

    /**
     * @param string $flavour (optional)
     *
     * @return string
     */
    public function getCommit($flavour = 'short')
    {
        switch ($flavour) {
            case 'full':
                exec('git log -1', $line);
                $hash = str_replace('commit ', '', $line[0]);
                break;
            default:
                exec('git describe --always', $line);
                $hash = array_shift($line);
                break;
        }
        $this->_sHeaderbarCommit = $hash;

        return $this->getStoreConfig(self::CONFIG_EXTENSION_HEADERBAR_COMMIT, '_sHeaderbarCommit');
    }

    /**
     * @return string
     */
    public function getClassName()
    {
        if (null !== $this->_className) {
            return $this->_className;
        }
        $this->_className = '';
        $domain = $this->getDomain();
        $environments = array(
            'production',
            'staging',
            'development',
        );
        foreach ($environments as $env) {
            if ($domain === Mage::getStoreConfig('ffuenf_devtools/headerbar/' . $env . '_path')) {
                $this->_className = $env;
                break;
            }
        }

        return $this->_className;
    }

    /**
     * @return string
     */
    public function getDomain()
    {
        if (null !== $this->_domain) {
            return $this->_domain;
        }
        $path = strtolower(rtrim(trim(Mage::getStoreConfig('web/unsecure/base_url')), '/'));
        $this->_domain = str_replace(array('http://', 'https://'), '', $path);
        return $this->_domain;
    }

    /**
     * @return boolean
     */
    public function supportsDesignFallback()
    {
        // Check for the right file
        return !(file_exists(BP . '/app/code/core/Mage/Core/Model/Design/Fallback.php') === false);
    }
    
    /**
     * Get timeframe of resaving products.
     *
     * @throws Mage_Core_Exception
     */
    public function getResaveProductsTimeframeFrom()
    {
        return strtotime($this->getStoreConfig(self::CONFIG_EXTENSION_RESAVEPRODUCTS_TIMEFRAMEFROM, '_sResaveProductsTimeframeFrom'));
    }

    /**
     * Get timeframe of resaving products.
     *
     * @throws Mage_Core_Exception
     */
    public function getResaveProductsTimeframeTo()
    {
        return strtotime($this->getStoreConfig(self::CONFIG_EXTENSION_RESAVEPRODUCTS_TIMEFRAMETO, '_sResaveProductsTimeframeTo'));
    }

    /**
     * Get timeframe type of resaving products.
     *
     * @throws Mage_Core_Exception
     * @return string
     */
    public function getResaveProductsTimeframeType()
    {
        return $this->getStoreConfig(self::CONFIG_EXTENSION_RESAVEPRODUCTS_TIMEFRAMETYPE, '_sResaveProductsTimeframeType');
    }

    public function isBirthday($birthdate)
    {
        if (empty($birthdate)) {
            return false;
        }
        $currentDayMonth = date("m-d");
        $birthdateDayMonth = substr($birthdate, 5, 5);
        return $currentDayMonth == $birthdateDayMonth;
    }

    public function getSeason()
    {
        if ($this->_isWinter()) {
            return 'winter';
        }
        if ($this->_isSpring()) {
            return 'spring';
        }
        if ($this->_isSummer()) {
            return 'summer';
        }
        if ($this->_isAutumn()) {
            return 'autumn';
        }
    }

    private function _isWinter()
    {
        $currentYear = date("Y");
        return $this->_isDateInRange("{$currentYear}-12-01", "{$currentYear}-12-31")
            || $this->_isDateInRange("{$currentYear}-01-01", "{$currentYear}-02-29");
    }

    private function _isSpring()
    {
        $currentYear = date("Y");
        return $this->_isDateInRange("{$currentYear}-03-01", "{$currentYear}-05-31");
    }

    private function _isSummer()
    {
        $currentYear = date("Y");
        return $this->_isDateInRange("{$currentYear}-06-01", "{$currentYear}-08-31");
    }

    private function _isAutumn()
    {
        $currentYear = date("Y");
        return $this->_isDateInRange("{$currentYear}-09-01", "{$currentYear}-11-30");
    }

    private function _isDateInRange($startDate, $endDate, $dateToCheck = null)
    {
        if (is_null($dateToCheck)) {
            $dateToCheck = time();
        }
        if (is_string($dateToCheck)) {
            $dateToCheck = strtotime($dateToCheck);
        }
        $startTimestamp = strtotime($startDate);
        $endTimestamp = strtotime($endDate);
        return (($dateToCheck >= $startTimestamp) && ($dateToCheck <= $endTimestamp));
    }
}
